### PR TITLE
Create opentelemetry transport

### PR DESCRIPTION
**Pull request by Vunovati**
Closes https://github.com/pinojs/pino/issues/1292

Implements Pino transport for OpenTelemetry. Outputs logs in the [OpenTelemetry Log Data Model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md).

The list of options for the transport is
* `destination`: The destination of the log.
* `messageKey`: The key of the log message to be used as the OpenTelemetry log entry Body. Optional, value `msg` used by default (like in Pino itself).

It is possible that some additional options could be useful but this is all that I figure is required for the basic use case.

## Test with
```
const pino = require('pino')

const transport = pino.transport({
  target: '.',
  options: { destination: 1 }
})

const logger = pino(transport)

transport.on('ready', () => {
  setInterval(() => {
    logger.info('test log')
  }, 1000)
})

```

### CODE REVIEW

**Review by rp4rk — APPROVED** [review]
LGTM, spotted a couple of typos and a question about the tests but other than that I can't spot anything 👍

File: .github/workflows/ci.yml
Comment (simoneb) [inline .github/workflows/ci.yml:25]: to reduce bumps frequency

```suggestion
        uses: actions/setup-node@v3
```

File: .github/workflows/ci.yml
Comment (simoneb) [inline .github/workflows/ci.yml:30]: there is no need for this action anymore, setup-node has built-in support for caching

File: .github/workflows/ci.yml
Comment (simoneb) [inline .github/workflows/ci.yml:36]: if you have a lockfile, run ci instead of i

```suggestion
        run: npm ci
```

File: package.json
Comment (simoneb) [inline package.json:13]: use your nearform email please

File: README.md
Comment (simoneb) [inline README.md:16]:
```suggestion
## Usage
```

**Review by simoneb — APPROVED** [review]
Good job 🚀 I'd try to get an approval from somebody in the pino core team for inclusion, but overall LGTM. I'm not too familiar with the protocol or any specific patterns being used in pino so it'd be good to have somebody else review t oo

File: test/pino-opentelemetry-transport.test.js
Comment (Vunovati) [reply test/pino-opentelemetry-transport.test.js]: Reduced to 100, lowering it would be too short (tests fail)

File: test/pino-opentelemetry-transport.test.js
Comment (simoneb) [inline test/pino-opentelemetry-transport.test.js:57]: isn't there a way to make this more robust, iow not relying on an arbitrary number but actually wait for something to happen?

Comment (Vunovati) [reply test/pino-opentelemetry-transport.test.js:57]: I cannot think of a simple solution for this. We could have a FS watcher or a periodic poller of sorts which would get notified on file changes and then wait for it to be updated 6 times. If we get past that watcher/poller that wait would already have confirmed what we are about to test in the remainder of the code. If the file does not get updated 6 times we would have to wait for some kind of timeout to pass.

I don't see this as a better option. Do you have some better idea? The overhead of the current solution is 300ms.

File: package-lock.json
Comment (jsumners) [inline package-lock.json:1]: We recommend not including this file in a library as it will not be used when the library is installed.

File: pino-opentelemetry-transport.js
Comment (jsumners) [inline pino-opentelemetry-transport.js:19]: Shouldn't this be a separate docblock? Wouldn't this `@typedef` conflict with the first `@typedef` in this block?

File: test/pino-opentelemetry-transport.test.js
Comment (jsumners) [reply test/pino-opentelemetry-transport.test.js:57]: Instead of writing to a file, write to a custom stream and inspect the lines. https://github.com/pinojs/pino/issues/1274#issuecomment-997971329

**Review by jsumners — APPROVED** [review]
This looks great to me.

File: pino-opentelemetry-transport.js
Comment (Vunovati) [reply pino-opentelemetry-transport.js:19]: This is in the same block to keep the related typedefs together.

It won't conflict: the Bindings extends the CommonBindings which is intentional.


The idea was to signify the difference between the usual Bindings from the unknown ones.

File: pino-opentelemetry-transport.js
Comment (Vunovati) [reply pino-opentelemetry-transport.js:49]: This was taken from the [Pino transports docs](https://github.com/pinojs/pino/blob/master/docs/transports.md#writing-a-transport).

File: pino-opentelemetry-transport.js
Comment (Vunovati) [reply pino-opentelemetry-transport.js:19]: In addition. The plan is to move everything to TypeScript soon so this will get deleted anyway.

Comment (jsumners) [reply pino-opentelemetry-transport.js:19]: 😬 that would result in me voting against adding it to the Pino organization. If that's not a goal, okay.

Comment (Vunovati) [reply pino-opentelemetry-transport.js:19]: Writing it in TypeScript? I assumed so, which was the reason why I originally wrote it in JS. 

I had a problem with generating types when using the original pino-abstract-transport so I had to [fork it](https://github.com/Vunovati/pino-abstract-transport?organization=Vunovati&organization=Vunovati). My fork changes how the types are exported enabling them to be picked up by TS compiler from JsDoc when generating the .d.ts file. I did not submit a PR since it did not seem worthwhile to disturb the waters over this. 

I had a talk outside of this PR with @simoneb where we concluded that it might be simpler to just use TS.

File: test/pino-opentelemetry-transport.test.js
Comment (Vunovati) [reply test/pino-opentelemetry-transport.test.js:57]: @jsumners, a question about the proposed solution. 
In [the code snippet](https://github.com/pinojs/pino/issues/1274#issuecomment-997971329) does the split function come from [split2](https://www.npmjs.com/package/split2) package?

In the opentelemetry transport the destination is being forwarded directly to the SonicBoom.
If the context.dest is a stream then SonicBoom cannot accept it.

`SonicBoom supports only file descriptors and files`

Comment (jsumners) [reply test/pino-opentelemetry-transport.test.js:57]: > In [the code snippet](https://github.com/pinojs/pino/issues/1274#issuecomment-997971329) does the split function come from [split2](https://www.npmjs.com/package/split2) package?

I believe so, yes. Unfortunately, I no longer have access to that code to verify.

> In the opentelemetry transport the destination is being forwarded directly to the SonicBoom.
> If the context.dest is a stream then SonicBoom cannot accept it.

I missed that. I think the transport should be able to write to a stream. This is the baseline functionality of a transport -- https://getpino.io/#/docs/transports?id=v7-transports

Comment (Vunovati) [reply test/pino-opentelemetry-transport.test.js:57]: Shouldn't the options such as dest be serializable? The code is executed in a worker thread.

This is specific to v7 transports I think.

_Note here we've specified a module by package rather than by relative path. The options object we provide is serialized and injected into the transport worker thread, then passed to the module's exported function. This means that the options object can only contain types that are supported by the [Structured Clone Algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) which is used to (de)serializing objects between threads._
[V7 transports docs](https://github.com/pinojs/pino/blob/master/docs/transports.md#v7-transports)

File: pino-opentelemetry-transport.js
Comment (Vunovati) [reply pino-opentelemetry-transport.js:19]: I updated the PR to include the manually generated types like in pino-abstract-transport. So no TypeScript needed.

Comment (mcollina) [reply pino-opentelemetry-transport.js:19]: Maybe fix pino-abstract-transport then?

File: test/pino-opentelemetry-transport.test.js
Comment (mcollina) [reply test/pino-opentelemetry-transport.test.js:57]: I don’t understand the question. Can you make it a bit clearer?

**Comment by mcollina** [comment]
I think this should send data to the OpenTelemetry collector and not to stdout, or maybe I did not understand how this would work: https://opentelemetry.io/docs/collector/.

**Comment by mcollina** [comment]
Anyway, this should be considered done with a demo that shows how to process this data in a opentelemetry-capable backend.

File: pino-opentelemetry-transport.js
Comment (mcollina) [inline pino-opentelemetry-transport.js:134]: Fastify and pino-http provides a `reqId` property that is unique for a given request. It might be nice to map that to a trace or a span.

We might want to have some logic to decide if an http request was completed. I'd be happy to add some property in the logs of Fastify/pino-http for this purpose or maybe we could just match the message text (and possibly make this configurable).

**Review by mcollina — CHANGES_REQUESTED** [review]

File: pino-opentelemetry-transport.js
Comment (Vunovati) [reply pino-opentelemetry-transport.js:19]: Sure, I have a fix ready. I'll create an issue on pino-abstract-trasport describing what the problem was and submit the PR.

**Comment by Vunovati** [comment]
> Anyway, this should be considered done with a demo that shows how to process this data in a opentelemetry-capable backend.

Makes sense. I'll create an example similar [to this](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/examples/demo/docker-compose.yaml)

File: pino-opentelemetry-transport.js
Comment (Vunovati) [reply pino-opentelemetry-transport.js:134]: Ok, I was thinking of adding a custom mapper object to this transport's options. The user could then decide how they want their props mapped. The mapper would override the default behaviour.

File: test/pino-opentelemetry-transport.test.js
Comment (Vunovati) [reply test/pino-opentelemetry-transport.test.js:57]: To clarify:

Is it a requirement for a pino transport to accept a stream as a parameter? Like this:

```
const transport = pino.transport({
  target: 'pino-opentelemetry-transport',
  options: { destination: someWriteStream }
})

```
Or is it ok for it to just accept filenames [like in the example from the docs](https://github.com/pinojs/pino/blob/master/docs/transports.md#writing-a-transport)?

The OT-transport implementation does not accept a stream but instantiates a SonicBoom stream which writes to the specified destination.

**Comment by Vunovati** [comment]
> I think this should send data to the OpenTelemetry collector and not to stdout, or maybe I did not understand how this would work: https://opentelemetry.io/docs/collector/.

My idea was to pipe the app stdout to a network client which would send the log lines to the collector. That approach seemed more unixy to me but I haven't tested it. I'll try to demo it in the example to see if it feasible. Maybe I am missing something 🤔

File: test/pino-opentelemetry-transport.test.js
Comment (mcollina) [reply test/pino-opentelemetry-transport.test.js:57]: Who will pick this data up and collect it? I'm totally ok this being a file but then who will deliver it? There is no requirement for this to accept a stream.

**Comment by jsumners** [comment]
It would make most sense if this module wrote to an OTL receiver -- https://github.com/open-telemetry/opentelemetry-collector/blob/95b976fabbccffc1d841fea55b668faf5be06970/receiver/otlpreceiver/README.md

Ideally, that would be a local agent that handles the shipping of stuff, but that would be up to the OTL implementation a user is using.

Falling back to writing to a log file seems like a good option to have, but I don't think it should be the primary.

**Comment by Vunovati** [comment]
Here is an update on the PR progress.

@simoneb I did not address some of the PR comments as ATM there is a bigger fish to fry, which could make some of the code obsolete. 

I've been researching how to implement sending to the OTLP collector as suggested by @jsumners. 

The ideal way to send the logs line-by-line to the OTLP collector would be to use the [opentelemetry-js sdk](https://github.com/open-telemetry/opentelemetry-js). Unfortunately, the support for logs is [not yet added](https://github.com/open-telemetry/opentelemetry-js#feature-status) (it is on the roadmap).

As a temporary workaround, we could send requests to the [OTLP Collector's OTLP/HTTP interface](https://github.com/open-telemetry/opentelemetry-collector/blob/95b976fabbccffc1d841fea55b668faf5be06970/receiver/otlpreceiver/README.md#writing-with-httpjson) manually which is probably [better](https://docs.lightstep.com/docs/send-otlp-over-http-to-lightstep#common-use-cases-for-otlphttp) than using the GRPC interface. 
I've been experimenting with the HTTP interface and found that it is not compatible with the [OTLP Data model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md) that I based my transport code one. None of the documented [example requests](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#example-log-records) works either. 

I tested by [booting the collector demo](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/examples/demo) + additionally exposing the HTTP interface on the collector.
I then sent the request with `curl -iv -H "Content-Type: application/json" http://localhost:4318/v1/logs -d @ot-data-model-example.json`


I reverse-engineered the model [from the collector code](https://github.com/open-telemetry/opentelemetry-collector/blob/8eb68f40028df449463c79367a538a2b0ee8d887/pdata/internal/data/protogen/logs/v1/logs.pb.go#L454), but it looks nothing like what's described in the aforementioned OT Data model. This 

```json
{
  "resourceLogs": [
    {
      "resource": {},
      "scopeLogs": [
        {
          "scope": {},
          "logRecords": [
            {
              "body": {
                "stringValue": "Test Pino log"
              },
              "observedTimeUnixNano": 1,
              "severityNumber": 1,
              "severityText": "TRACE",
              "traceId": "5982fe77008310cc80f1da5e10147519"
            }
          ]
        }
      ]
    }
  ]
}
```

There are two possible courses of action that I am considering:
1. Continue with the reverse-engineered approach and hope that the discovered API won't change
2. Wait for the OT JS SDK to implement the support for logging API and then create the transport to use that

In theory, there is an option to log into a file and then and then have OpenTelemetry read from that file but I am not sure how that should work as [the documentation](https://opentelemetry.io/docs/reference/specification/logs/logging-library-sdk/#logging-to-file) is not very clear ATM.

**Comment by simoneb** [comment]
Thanks for the detailed update @Vunovati! As discussed previously, I am somewhat surprised at the lack of SDK support and the inconsistency of the specification and the implementation. Although it would be great to have an OTLP support built into pino, if we have nothing stable to implement against I would challenge the value of doing this work now.

Is there a way for us to keep an eye on the progress of the support for logs into the JS SDK and delay this work until then? For the time being I would be happy to go with the initial implementation that writes logs to the file system or stdout in a compatible format, but delay the sending to a collector until spec and sdk are stable and available.

WDYT?

**Comment by Vunovati** [comment]
> Thanks for the detailed update @Vunovati! As discussed previously, I am somewhat surprised at the lack of SDK support and the inconsistency of the specification and the implementation. Although it would be great to have an OTLP support built into pino, if we have nothing stable to implement against I would challenge the value of doing this work now.
> 
> Is there a way for us to keep an eye on the progress of the support for logs into the JS SDK and delay this work until then? For the time being I would be happy to go with the initial implementation that writes logs to the file system or stdout in a compatible format, but delay the sending to a collector until spec and SDK are stable and available.
> 
> WDYT?

I think there is no reason to merge this PR until we are 100% sure we can deliver what we promise. As previously suggested, there needs to be a demo in the Readme which can prove this transport is doing its job.

I started asking around the OpenTelemetry community for some cues on how to go forward with this and which approach to take. I will not continue with development until I find a definitive answer. I will update this conversation when I gain some new insights.

**Comment by simoneb** [comment]
Sounds good to me, it looks like the JS ecosystem of OTLP has not caught up yet, so no need to push it further than we can reasonably go with this.

**Comment by jsumners** [comment]
> I started asking around the OpenTelemetry community for some cues on how to go forward with this and which approach to take. I will not continue with development until I find a definitive answer. I will update this conversation when I gain some new insights.

I think that is the correct approach. It also sounds like a really good opportunity for someone that has a strong interest in this area to help inform and build out the OTLP spec.

**Comment by simoneb** [comment]
@jsumners (or anybody else), do we have any idea why the official SDK doesn't support logs yet? I mean it sounds like something fairly basic.

**Comment by jsumners** [comment]
> @jsumners (or anybody else), do we have any idea why the official SDK doesn't support logs yet? I mean it sounds like something fairly basic.

I really have not been following the upstream project. I thought it was much further along, though.

**Comment by simoneb** [comment]
> I thought it was much further along, though.

Same here. From their repo I can see there's a super organized governance around the project (many versions of the spec, supporting Node all the way back to 10) but then it's missing basic features. I followed up to the discussion opened by @Vunovati https://github.com/open-telemetry/opentelemetry-js/discussions/2915 to check if there's a way we can help with that.

**Comment by weyert** [comment]
The `otel-js` is implementing the logging specification and I am wondering if it would make sense to write a transporter for Pino that calls the underlying API of `otel-js`?

[1] https://github.com/open-telemetry/opentelemetry-js/tree/main/experimental/packages/api-logs
[2] https://github.com/open-telemetry/opentelemetry-js/pull/3549

**Comment by Vunovati** [comment]
> The `otel-js` is implementing the logging specification and I am wondering if it would make sense to write a transporter for Pino that calls the underlying API of `otel-js`?
> 
> [1] https://github.com/open-telemetry/opentelemetry-js/tree/main/experimental/packages/api-logs [2] [open-telemetry/opentelemetry-js#3549](https://github.com/open-telemetry/opentelemetry-js/pull/3549)

Thanks for the update. This seems like what we've been waiting for. I'll need to take a better look these days and reimplement this PR using the JS logs API.

**Comment by Vunovati** [comment]
I implemented sending logs to the OTLP collector using the SDK. Sorry @weyert, couldn't be done sooner as even though the [logs SDK PR](https://github.com/open-telemetry/opentelemetry-js/pull/3549) was merged, the necessary packages were still not published [until 8 days ago](https://www.npmjs.com/package/@opentelemetry/exporter-logs-otlp-grpc?activeTab=versions).

Testing this is tricky, so I had to disable unit tests for now. I will write a proper integration test against the OTLP collector instance similar to the example I just added to the readme. After that, I'll ask for another review and, hopefully, publish this soon.

**Comment by weyert** [comment]
@Vunovati Thank you! I am wondering if it makes sense to make the use of the OTLP collector exporter optional and just call the logs API in this transporter and move the responsibility (as opt-in) of actually exporting to the user's OpenTelemetry set up?

As my idea was to use pino for logging, backed by the Logs API and then have my general OpenTelemetry SDK configuration to set up tracing, metrics and logging. Where Metrics goes to Prometheus, Tracing and Logs to Google Cloud Operations.

**Comment by 10xLaCroixDrinker** [comment]
@Vunovati are you still working on this?

**Comment by Vunovati** [comment]
Yes, I am. Wanted to increase the test coverage before submitting the PR
for another review. Will push tomorrow so you can track the progress.

**Comment by 10xLaCroixDrinker** [comment]
@Vunovati What is your expected completion date? Trying to evaluate if I should wait for it or plan on implementing some other way.

**Comment by Vunovati** [comment]
> @Vunovati What is your expected completion date? Trying to evaluate if I should wait for it or plan on implementing it some other way.

The implementation is usable enough, so I'll merge the plugin this week the way it is. @jsumners, the Readme now contains the instructions on sending the logs to the OTLP collector as you requested. Also, this is covered by tests that can (only) be run locally with `npm test`. The tests are read from the log file created by the OTLP collector. Can you please take another look? If you think this is good enough to get accepted to the Pino organization. If not, I'll create a new PR addressing any work left to get it accepted.

File: README.md
Comment (jsumners) [inline README.md:19]: Punctuation should be added, and I think the noted variable are environment variables? That should be made clear.

File: README.md
Comment (jsumners) [inline README.md:1]: Make sure all sentences have punctuation.

File: otlp-logger.js
Comment (jsumners) [inline otlp-logger.js:68]: Use a permanent link.

File: otlp-logger.js
Comment (jsumners) [inline otlp-logger.js:79]: Use a permanent link.

File: package-lock.json
Comment (jsumners) [inline package-lock.json:1]: This file should be removed and excluded from git.

File: package.json
Comment (jsumners) [inline package.json:27]: Consider using https://github.com/fastify/pre-commit instead.

**Review by jsumners — COMMENTED** [review]
Looking pretty good. I can't speak to any of the details. I assume you've covered what needs to be covered.

If it is added to the pinojs org, would you still be able to help maintain it?

**Comment by Vunovati** [comment]
> Looking pretty good. I can't speak to any of the details. I assume you've covered what needs to be covered.
> 
> If it is added to the pinojs org, would you still be able to help maintain it?

Yes, I would. The module is small enough.

**Comment by jsumners** [comment]
ping: @mcollina

File: otlp-logger.js
Comment (10xLaCroixDrinker) [inline otlp-logger.js:31]:
```suggestion
      'service.version': opts.serviceVersion,
      ...opts.resourceAttributes
```

Can you add an option for additional resource attributes? I feel this is a must-have

File: otlp-logger.js
Comment (10xLaCroixDrinker) [inline otlp-logger.js:143]: Is this needed? OTel will add a timestamp

File: otlp-logger.js
Comment (10xLaCroixDrinker) [inline otlp-logger.js:105]:
```suggestion
const SEVERITY_NAME_MAP = Object.entries(SeverityNumber)
  .reduce((acc, [severityText, severityNumber]) => ({
    ...acc,
    [severityNumber]: severityText,
  }), {});
```

Can this be done programmatically to avoid errors and keep up with future changes or additions? Currently this is missing UNSPECIFIED

File: otlp-logger.js
Comment (10xLaCroixDrinker) [inline otlp-logger.js:139]: Why default to fatal?

File: otlp-logger.js
Comment (10xLaCroixDrinker) [inline otlp-logger.js:10]:
```suggestion
const { SeverityNumber, logs } = require('@opentelemetry/api-logs')
```

File: otlp-logger.js
Comment (10xLaCroixDrinker) [inline otlp-logger.js:40]: What if someone wants multiple exporters? e.g. `ConsoleLogRecordExporter` & `OTLPLogExporter`

**Review by 10xLaCroixDrinker — CHANGES_REQUESTED** [review]

File: otlp-logger.js
Comment (Vunovati) [reply otlp-logger.js:143]: In the resulting logs, There are two timestamps:

1. timeUnixNano - one that is provided by the user and specified in the data model as [Timestamp](https://github.com/open-telemetry/opentelemetry-specification/blob/fc8289b8879f3a37e1eba5b4e445c94e74b20359/specification/logs/data-model.md#field-timestamp)
2. observedTimeUnixNano - one that is added by OTLP SDK and specified in the data model as [ObservedTimestamp](https://github.com/open-telemetry/opentelemetry-specification/blob/fc8289b8879f3a37e1eba5b4e445c94e74b20359/specification/logs/data-model.md#field-observedtimestamp)


in 1. we are sending the time as observed by Pino.

File: otlp-logger.js
Comment (Vunovati) [reply otlp-logger.js:105]: Can you provide the link to the spec mentioning UNSPECIFIED?

File: otlp-logger.js
Comment (Vunovati) [reply otlp-logger.js:139]: What would you suggest as an alternative? My line of thinking was, I definitely did not want to throw an error in a logger. OTOH, If my logging is not working, I want people to get alarmed. Is there something in the OTLP spec that I am missing?

File: otlp-logger.js
Comment (Vunovati) [reply otlp-logger.js:40]: That's a good point. I couldn't think of a simple API that would support various combinations of exporters (GRPC, HTTP, Console, InMemory ...) and processors(Batch, simple ...). Do you have some suggestions? I'll create a dedicated issue to discuss and address this.

File: otlp-logger.js
Comment (10xLaCroixDrinker) [reply otlp-logger.js:30]: It's my experience with the OTel node SDK is that this environment variable isn't used automatically. I've had to parse it myself and pass it in. Does yours differ?

File: otlp-logger.js
Comment (10xLaCroixDrinker) [reply otlp-logger.js:142]: The SDK adds a timestamp in the unix nano format. The observed timestamp is added by the collector, not the SDK

File: otlp-logger.js
Comment (10xLaCroixDrinker) [reply otlp-logger.js:138]: In OTel implementation, I use `UNSPECIFIED` (`0`) when there is no severity specified.

File: otlp-logger.js
Comment (10xLaCroixDrinker) [reply otlp-logger.js:30]: There are also attributes that you may want to add that would not be best set in an environment variable. e.g...

```js
{ [SemanticResourceAttributes.SERVICE_INSTANCE_ID]: `${os.hostname()}:${process.pid}` }
```

File: otlp-logger.js
Comment (Vunovati) [reply otlp-logger.js]: Can you clarify what the problem is with having both fields?h The [docs for the field](https://github.com/open-telemetry/opentelemetry-specification/blob/fc8289b8879f3a37e1eba5b4e445c94e74b20359/specification/logs/data-model.md#field-timestamp) say:
_This field is optional, it may be missing if the source timestamp is unknown._

in our case, the source timestamp is known, so why not set it? Each has its use and means different things. If you run the example from the Readme, you will get something like the following logRecords (in `/tmp/test-logs/otlp-logs.log`)



```
          "logRecords": [
            {
              "timeUnixNano": "1690326690783000064",
              "observedTimeUnixNano": "1690326690964999936",
...
            },
            {
              "timeUnixNano": "1690326691784000000",
              "observedTimeUnixNano": "1690326691996000000",
...
            },
            {
              "timeUnixNano": "1690326692784000000",
              "observedTimeUnixNano": "1690326693028999936",
...
            },
            {
              "timeUnixNano": "1690326693784999936",
              "observedTimeUnixNano": "1690326694059000064",
...
            },
            {
              "timeUnixNano": "1690326694785999872",
              "observedTimeUnixNano": "1690326695091000064",
...
            }


```
I assume one signifies the pino logger's time, while the other signifies the time when the batch containing the log record was sent.

This is how the values get set initially:

https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/sdk-logs/src/LogRecord.ts#L88

So the behaviour you recommend is a fallback behaviour. I'd rather go with capturing more info in the log.

File: otlp-logger.js
Comment (Vunovati) [reply otlp-logger.js:30]: > It's my experience with the OTel node SDK is that this environment variable isn't used automatically. I've had to parse it myself and pass it in. Does yours differ?

It is possible to do that. You are correct this does not work in the current implementation. I should add a ResourceDetector to read the attributes.

> There are also attributes that you may want to add that would not be best set in an environment variable. e.g...
> 
> ```js
> { [SemanticResourceAttributes.SERVICE_INSTANCE_ID]: `${os.hostname()}:${process.pid}` }
> ```

Good point. We should expand the API for that kind of use cases.

Comment (Vunovati) [reply otlp-logger.js:30]: https://github.com/Vunovati/pino-opentelemetry-transport/issues/4

File: otlp-logger.js
Comment (Vunovati) [reply otlp-logger.js:39]: https://github.com/Vunovati/pino-opentelemetry-transport/issues/3

File: otlp-logger.js
Comment (Vunovati) [reply otlp-logger.js]: Thanks for the suggestions, If I missed something here, please create an Issue, and we can continue there. I'll merge the PR so that we do not expand the scope of this PR indefinitely

**Pull request merged**
